File: desktop/src/shared/api/relayRateLimitGate.ts
/**
 * Module-level rate-limit gate for relay WebSocket operations.
 *
 * WebSocket back-pressure activates this gate. HTTP 429s are handled by the
 * native HTTP gate: ApiCalls and WsEvents are separate relay quota budgets.
 * Callers that must not run while WS-limited await `waitForRateLimit()`.
 *
 * The gate is a singleton: one shared expiry covers all concurrent callers so
 * overlapping hints (multiple CLOSED frames) extend to the latest expiry without
 * stacking. This module must be reset on community switch via
 * `resetRateLimitGate()`.
 */

/** Minimum gate duration when the relay provides no `retry in Ns` hint. */
const DEFAULT_RATE_LIMIT_SECONDS = 10;

/**
 * Maximum hint the TS gate will honour from a WebSocket rejection.
 *
 * Uses the same cap as the native HTTP gate (`MAX_HINT_SECONDS` in
 * `relay_admission.rs`), but each transport consumes its own retry hints.
 */
export const MAX_HINT_SECONDS = 300;

let expiresAt: number | null = null;
let gateTimer: number | null = null;
let gateResolve: (() => void) | null = null;
let gatePromise: Promise<void> | null = null;

/**
 * Parse `retry in Ns` from a relay message string.
 *
 * Matches the relay's canonical hint format embedded in both CLOSED messages
 * (`rate-limited: quota exceeded; retry in 4s`) and HTTP 429 bodies
 * (`relay rate-limited: retry in 4s`).
 */
export function parseRateLimitHint(msg: string): number | null {
  const match = /retry in (\d+)s/i.exec(msg);
  return match ? parseInt(match[1], 10) : null;
}

/**
 * Activate (or extend) the rate-limit gate.
 *
 * If the gate is already active, the expiry is pushed forward to the maximum of
 * the existing expiry and the new hint — overlapping hints never shrink the
 * window. Non-positive or absent hints use the 10-second default; a 0s gate
 * would resolve immediately and swallow the signal.
 *
 * Note: buzz-acp uses a 5s no-hint default; desktop deliberately uses 10s here
 * for a wider back-off window on degraded connections.
 */
export function activateRateLimit(retryInSeconds: number | null): void {
  const durationMs =
    (retryInSeconds != null && retryInSeconds > 0
      ? Math.min(retryInSeconds, MAX_HINT_SECONDS)
      : DEFAULT_RATE_LIMIT_SECONDS) * 1_000;
  const newExpiry = Date.now() + durationMs;

  if (expiresAt !== null && newExpiry <= expiresAt) {
    // Existing window already covers this hint — nothing to do.
    return;
  }

  expiresAt = newExpiry;

  if (gateTimer !== null) {
    window.clearTimeout(gateTimer);
  }

  if (gatePromise === null) {
    // First activation: create the shared promise that waiters will await.
    gatePromise = new Promise<void>((resolve) => {
      gateResolve = resolve;
    });
  }

  gateTimer = window.setTimeout(() => {
    gateTimer = null;
    expiresAt = null;
    const resolve = gateResolve;
    gateResolve = null;
    gatePromise = null;
    resolve?.();
  }, durationMs);
}

/**
 * Arms the gate if `message` is a relay back-pressure signal, and reports
 * whether it was.
 *
 * The relay marks back-pressure with a `rate-limited:` prefix on whichever
 * frame carries the rejection — `NOTICE` for connection-scoped limits, `OK`
 * for one addressed to a single event, `CLOSED` for a subscription. Every
 * inbound path needs the same test, so it lives here with the gate rather than
 * being re-derived per call site.
 */
export function activateRateLimitIfSignalled(message: string): boolean {
  if (!message.startsWith("rate-limited:")) {
    return false;
  }
  activateRateLimit(parseRateLimitHint(message));
  return true;
}

/** Returns `true` when the relay has signalled back-pressure and the gate is active. */
export function isRateLimited(): boolean {
  return expiresAt !== null && Date.now() < expiresAt;
}

/**
 * Resolves when the rate-limit gate expires.
 *
 * Resolves immediately if the gate is not active. Callers that need to wait
 * before issuing a new relay request should await this before proceeding.
 */
export function waitForRateLimit(): Promise<void> {
  if (!isRateLimited() || gatePromise === null) {
    return Promise.resolve();
  }
  return gatePromise;
}

/**
 * Returns the milliseconds remaining on the active gate, or 0 when inactive.
 *
 * Use this instead of re-deriving the hint from the message so that a shorter
 * relay hint arriving under a longer active gate never schedules a premature
 * retry.
 */
export function rateLimitRemainingMs(): number {
  if (expiresAt === null) return 0;
  return Math.max(0, expiresAt - Date.now());
}

/**
 * Reset all gate state. Must be called on community switch so a rate-limit hint
 * from the old relay does not bleed into the new community's session.
 *
 * Any in-flight `waitForRateLimit()` awaiters are resolved immediately so they
 * do not leak into the new session.
 */
export function resetRateLimitGate(): void {
  if (gateTimer !== null) {
    window.clearTimeout(gateTimer);
    gateTimer = null;
  }
  expiresAt = null;
  const resolve = gateResolve;
  gateResolve = null;
  gatePromise = null;
  resolve?.();
}

File: mobile/lib/shared/relay/relay_rate_limit_gate.dart
import 'dart:async';
import 'dart:math';

/// Creates a timer used by [RelayRateLimitGate].
typedef RelayTimerFactory =
    Timer Function(Duration duration, void Function() callback);

/// Session-owned gate that pauses relay requests after back-pressure.
class RelayRateLimitGate {
  /// Default gate duration when the relay omits a parseable retry hint.
  static const defaultRetrySeconds = 10;

  /// Longest retry hint accepted from a relay response.
  static const maxRetrySeconds = 300;

  RelayRateLimitGate({
    DateTime Function()? now,
    RelayTimerFactory timerFactory = Timer.new,
  }) : _now = now ?? DateTime.now,
       _timerFactory = timerFactory;

  final DateTime Function() _now;
  final RelayTimerFactory _timerFactory;
  DateTime? _expiresAt;
  Timer? _timer;
  Completer<void>? _completer;

  /// Whether a rate-limit window is currently active.
  bool get isActive {
    final expiresAt = _expiresAt;
    return expiresAt != null && _now().isBefore(expiresAt);
  }

  /// Activates or extends the gate without shrinking an existing window.
  ///
  /// A null [retryInSeconds] means the relay sent no parseable hint, so the
  /// conservative [defaultRetrySeconds] applies. An explicit hint is honored as
  /// given: the relay sends `retry in 0s` to mean "retry immediately", so a
  /// non-positive hint opens no window at all. Treating `0` as if it were
  /// absent previously gated every relay read for [defaultRetrySeconds] — on a
  /// cold start with many channels that turned the relay's own "go now" into a
  /// ten-second stall of the whole session.
  ///
  /// An already-active window is never shortened, so an immediate hint arriving
  /// mid-window leaves that window intact.
  void activate(int? retryInSeconds) {
    if (retryInSeconds != null && retryInSeconds <= 0) return;
    final seconds = retryInSeconds == null
        ? defaultRetrySeconds
        : min(retryInSeconds, maxRetrySeconds);
    final duration = Duration(seconds: seconds);
    final newExpiry = _now().add(duration);
    final currentExpiry = _expiresAt;
    if (currentExpiry != null && !newExpiry.isAfter(currentExpiry)) return;

    _expiresAt = newExpiry;
    _timer?.cancel();
    _completer ??= Completer<void>();
    _timer = _timerFactory(duration, _expire);
  }

  /// Resolves when the active rate-limit window expires.
  Future<void> wait() {
    if (!isActive) return Future.value();
    return _completer!.future;
  }

  /// Milliseconds remaining in the active window, or zero when inactive.
  int remainingMs() {
    final expiresAt = _expiresAt;
    if (expiresAt == null) return 0;
    return max(0, expiresAt.difference(_now()).inMilliseconds);
  }

  /// Clears the gate and releases all current waiters.
  void reset() {
    _timer?.cancel();
    _timer = null;
    _expiresAt = null;
    final completer = _completer;
    _completer = null;
    if (completer != null && !completer.isCompleted) completer.complete();
  }

  void _expire() {
    _timer = null;
    _expiresAt = null;
    final completer = _completer;
    _completer = null;
    if (completer != null && !completer.isCompleted) completer.complete();
  }
}
